pkg/server/http/ipfs.go

```diff
@@ -0,0 +1,184 @@
+		res.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	// check if cid path param is missing
+	if len(urlPath) < 2 {
+		// not a valid path to hit
+		logger.LogStatus(http.StatusNotFound, "Not found")
+		res.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	// check if Accept header includes application/vnd.ipld.car
+	hasAccept := req.Header.Get("Accept") != ""
+	acceptTypes := strings.Split(req.Header.Get("Accept"), ",")
+	validAccept := false
+	for _, acceptType := range acceptTypes {
+		typeParts := strings.Split(acceptType, ";")
+		if typeParts[0] == "*/*" || typeParts[0] == "application/*" || typeParts[0] == "application/vnd.ipld.car" {
+			validAccept = true
+			break
+		}
+	}
+	if hasAccept && !validAccept {
+		logger.LogStatus(http.StatusBadRequest, "No acceptable content type")
+		res.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// check if format is car
+	hasFormat := req.URL.Query().Has("format")
+	if hasFormat && req.URL.Query().Get("format") != "car" {
+		logger.LogStatus(http.StatusBadRequest, fmt.Sprintf("Requested non-supported format %s", req.URL.Query().Get("format")))
+		res.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// if neither are provided return
+	// one of them has to be given with a car type since we only return car files
+	if !validAccept && !hasFormat {
+		logger.LogStatus(http.StatusBadRequest, "Neither a valid accept header or format parameter were provided")
+		res.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// check if provided filename query parameter has .car extension
+	if req.URL.Query().Has("filename") {
+		filename := req.URL.Query().Get("filename")
+		ext := filepath.Ext(filename)
+		if ext == "" {
+			logger.LogStatus(http.StatusBadRequest, "Filename missing extension")
+			res.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		if ext != ".car" {
+			logger.LogStatus(http.StatusBadRequest, fmt.Sprintf("Filename uses non-supported extension %s", ext))
+			res.WriteHeader(http.StatusBadRequest)
+			return
+		}
+	}
+
+	// validate cid path parameter
+	cidStr := urlPath[1]
+	cid, err := cid.Parse(cidStr)
+	if err != nil {
+		logger.LogStatus(http.StatusInternalServerError, "Failed to parse cid path parameter")
+		res.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Grab unixfs path if it exists
+	// var unixfsPath []string
+	// if len(urlPath) > 2 {
+	// 	unixfsPath = urlPath[2:]
+	// }
+	// TODO: Do something with unixfs path
+
+	log.Debug("creating temp car file")
+	carfile, err := os.CreateTemp("", cid.String())
+	if err != nil {
+		logger.LogStatus(http.StatusInternalServerError, fmt.Sprintf("Failed to create temp car file before retrieval: %v", err))
+		res.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	var lassie = lassie.NewLassie(20 * time.Second)
+	log.Debugw("fetching cid into car file", "cid", cid.String(), "file", carfile.Name())
+	_, id, err := lassie.Fetch(context.Background(), cid, carfile)
```

Comment:
```suggestion
	err, id, err := lassie.Fetch(req.Context(), cid, carfile)
```

error not handled, and `req.Context()` should give us client-cancellation flowing through the entire retrieval process

Reply:
Ah! I was not aware request had a context. Good suggestion, thank you.